File: json_submit.py
import csv
import datetime
import json
import os
import re

from dateutil import parser
from pyspark import SparkContext
from pyspark.sql import SparkSession
from pyspark.sql import functions as F
from pyspark.sql import types as T

try:
    spark
except NameError:
    spark = SparkSession.builder.appName("proj").getOrCreate()


def task1():
    ret = {}
    for filename in os.listdir('/scratch/ql1045/'):
        if filename.startswith('task1.'):
            with open(f'/scratch/ql1045/{filename}') as f:
                parsed = json.load(f)
                for column in j['columns']:
                    for data_type in list(column['data_types']):
                        if data_type['type'] == 'INTEGER (LONG)':
                            int_cap = data_type['count']
                        elif data_type['type'] == 'REAL':
                            double_cap = data_type['count']
                            if double_cap == int_cap:
                                column['data_types'].remove(data_type)
                        elif data_type['type'] == 'DATE/TIME':
                            date_cap = data_type['count']
                            if date_cap == int_cap or date_cap == double_cap:
                                column['data_types'].remove(data_type)
                ret[re.search(r'[\d\w]{4}[-][\d\w]{4}', filename).group(0)] = parsed
    json.dump(ret, f, indent=2)


def task2_manual_label():
    ground_truth = {
        filename: [
            # label.filter(e => e)
            semantic_type
            for semantic_type in labels if semantic_type
        ]
        # for each line
        for [filename, *labels] in (spark.read.format('csv')
                                    .options(header='true', inferschema='true')
                                    .load('/user/ql1045/proj-in/DF_Label.csv')
                                    .toLocalIterator())
    }
    with open('task2-manual-labels.json', 'w') as f:
        json.dump(ground_truth, f, indent=2)

File: json-submit.js
const fs = require('fs');
const JSON5 = require('json5')

// task1.json
var jsons = fs.readdirSync('.').filter(e => e.includes('task1.')).map(e => JSON5.parse(fs.readFileSync(e)));
var ret = { datasets: jsons };
for (const { columns } of ret.datasets) {
    for (const { data_types } of columns) {
        const { length } = data_types;
        if (length > 1) { }
        const int_cap = (data_types.find(({ type }) => type == 'INTEGER (LONG)') || {}).count;
        const real_cap = (data_types.find(({ type }) => type == 'REAL') || {}).count;
        const date_cap = (data_types.find(({ type }) => type == 'DATE/TIME') || {}).count;
        if (real_cap && real_cap === int_cap)
            data_types.splice(data_types.findIndex(({ type }) => type == 'REAL'));
        if (date_cap && (date_cap === int_cap || date_cap === real_cap))
            data_types.splice(data_types.findIndex(({ type }) => type == 'DATE/TIME'));
    }
}
fs.writeFileSync('task1.json', JSON.stringify(ret, null, 2));

// task1 stat
var stat = {}
for (const { columns } of ret.datasets) {
    for (const { data_types } of columns) {
        for (const { type } of data_types) {
            stat[type] = (stat[type] || 0) + 1;
        }
    }
}


// task2.json
var jsons = fs.readdirSync('task2').filter(e => e.includes('.json')).map(e => JSON5.parse(fs.readFileSync(e)));


// task2-manual-labels.json
var csv = fs.readFileSync('./DF_Label.csv', 'utf8').split('\n').map(e => e.split(',').map(e => e.trim()).filter(e => e));
var csv = csv.slice(1);
var ret = {
    actual_types: csv.map(([column_name, ...labels]) => ({
        column_name,
        manual_labels: labels.map(semantic_type => ({ semantic_type })),
    }))
}
fs.writeFileSync('task2-manual-labels.json', JSON.stringify(ret, null, 2));

// 2e
var [correct_postive, predict_postive, truth_postive] = a.split('\n').map(e => JSON.parse(e))
console.log('precision')
Object.fromEntries(Object.entries(correct_postive).map(([k, v]) => [k, v / predict_postive[k]]))
console.log('recall')
Object.fromEntries(Object.entries(correct_postive).map(([k, v]) => [k, v / truth_postive[k]]))
